### PR TITLE
[substack] Add new extractor

### PR DESCRIPTION
**Pull request by elyse0**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

> Note: This is a follow-up of #3804 after a bad rebase

**Description**

This PR adds a new extractor for podcast episodes and videos inside https://substack.com/ and also custom domains

Fixes https://github.com/yt-dlp/yt-dlp/issues/3722

### Custom domains

```bash
yt-dlp -vU "https://www.mollymovieclub.com/p/interstellar?s=r#details"
```

```bash
[debug] Command-line config: ['-vU', 'https://www.mollymovieclub.com/p/interstellar?s=r#details']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.05.18 [b14d52355] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: 3aa741f89
[debug] Python version 3.8.10 (CPython 64bit) - Linux-5.4.0-113-generic-x86_64-with-glibc2.29
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg N-106673-g058a1ff9b4-20220424 (setts), ffprobe N-106673-g058a1ff9b4-20220424, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
Latest version: 2022.05.18, Current version: 2022.05.18
yt-dlp is up to date (2022.05.18)
[debug] [generic] Extracting URL: https://www.mollymovieclub.com/p/interstellar?s=r#details
[generic] interstellar?s=r#details: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] interstellar?s=r#details: Downloading webpage
[generic] interstellar?s=r#details: Extracting information
[debug] Looking for video embeds
[debug] [Substack] Extracting URL: https://mollymovieclub.substack.com/p/interstellar?s=r#details
[Substack] interstellar: Downloading webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 53602801: Chapter information is unavailable
[info] 53602801: Downloading 1 format(s): 0
[debug] Invoking http downloader on "https://substackcdn.com/public/audio/e151e11d-313b-4038-a3df-b0a87bf99ff8.mpga?post_id=53602801"
[download] Resuming download at byte 4193280
[download] Destination: Interstellar [53602801].mpga
[download]   7.2% of 165.92MiB at    4.33MiB/s ETA 00:35
```

### Substack blogs

```bash
[debug] Command-line config: ['-vU', 'https://haleynahman.substack.com/p/i-made-a-vlog?s=r']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.05.18 [b14d52355] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: 3aa741f89
[debug] Python version 3.8.10 (CPython 64bit) - Linux-5.4.0-113-generic-x86_64-with-glibc2.29
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg N-106673-g058a1ff9b4-20220424 (setts), ffprobe N-106673-g058a1ff9b4-20220424, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
Latest version: 2022.05.18, Current version: 2022.05.18
yt-dlp is up to date (2022.05.18)
[debug] [Substack] Extracting URL: https://haleynahman.substack.com/p/i-made-a-vlog?s=r
[Substack] i-made-a-vlog: Downloading webpage
[Substack] a7315fa8-8027-4432-a7e3-652479bc5275: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 47660949: Chapter information is unavailable
[info] 47660949: Downloading 1 format(s): 1
[debug] Invoking http downloader on "https://haleynahman.substack.com/api/v1/video/upload/a7315fa8-8027-4432-a7e3-652479bc5275/src?type=mp4"
[download] Destination: I MADE A VLOG [47660949].mp4
[download]   0.1% of 1.13GiB at    2.44MiB/s ETA 07:53
```

### CODE REVIEW

**Comment by elyse0** [comment]
> Pukkandan: Are you working on https://github.com/yt-dlp/yt-dlp/pull/3804#issuecomment-1138077801 or should I merge without it?

Yes, I've just added support for custom domains

File: yt_dlp/extractor/generic.py [hunk 1]
```diff
@@ -3037,6 +3062,10 @@ def _real_extract(self, url):
             # Don't set the extractor because it can be a track url or an album
             return self.url_result(burl)
 
+        # Check for Substack custom domains
+        if SubstackIE.is_custom_domain(webpage):
+            return self.url_result(SubstackIE.url_from_custom_domain(url))
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3067]: use `_extract_url` like other extractors

File: yt_dlp/extractor/substack.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                })
+
+        return formats, subtitles
+
+    def _real_extract(self, url):
+        display_id, username = self._match_valid_url(url).group('id', 'username')
+        webpage = self._download_webpage(url, display_id)
+
+        post_info = self._search_json(
+            r'<script[^>]*>\s*window\._preloads\s*=\s*', webpage, 'preloads', display_id)['post']
```
Comment (pukkandan) [inline yt_dlp/extractor/substack.py:78]:
```suggestion
            r'<script[^>]*>\s*window\._preloads\s*=', webpage, 'preloads', display_id)['post']
```

File: yt_dlp/extractor/substack.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        return re.search(r'<script[^>]+src=["\']https://substackcdn.com/.*\.js', webpage) is not None
+
+    @staticmethod
+    def url_from_custom_domain(url):
+        custom_domain = r'https?://(?:www\.)?(?P<username>[\w-]+)\.(?:[a-z]+)/(?P<params>.+)'
+
+        username = re.search(custom_domain, url).group('username')
+        params = re.search(custom_domain, url).group('params')
+
+        return f'https://{username}.substack.com/{params}'
```
Comment (pukkandan) [inline yt_dlp/extractor/substack.py:54]: like this

```suggestion
    @classmethod
    def _extract_url(cls, webpage, url):
        if not re.search(r'<script[^>]+src=["\']https://substackcdn.com/[^"\']+\.js', webpage):
            return
        parsed = urlparse(url)
        return parsed._replace(netloc=f'{parsed.netloc.split(".")[-2]}.substack.com').geturl()
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/substack.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+        self._sort_formats(formats)
+        return {
+            'id': str(post_info['id']),
+            'formats': formats,
+            'subtitles': subtitles,
+            'title': post_info.get('title'),
+            'description': post_info.get('description'),
+            'thumbnail': post_info.get('cover_image'),
+            'uploader': username,
```
Comment (fstirlitz) [inline yt_dlp/extractor/substack.py:97]: `uploader` should probably somehow point to the Substack profile page (under `https://substack.com/profile/`). What you call `username` here is actually the name of the newsletter only.

Compare mollymovieclub.com and computerenhance.substack.com being both written by the same author.

File: yt_dlp/extractor/substack.py (hunk 3 above)
Comment (fstirlitz) [reply yt_dlp/extractor/substack.py:54]: Do we actually have a guarantee that a Substack custom domain corresponds to an identically-spelt subdomain of `substack.com`?

Comment (elyse0) [reply yt_dlp/extractor/substack.py]: I just noticed we can get the subdomain from the page, can you check this commit? https://github.com/yt-dlp/yt-dlp/pull/4011/commits/2a563e100949122ea6a8c40a73c6138ddbab3cd2

Comment (fstirlitz) [reply yt_dlp/extractor/substack.py]: Okay, I like that more. Personally, I would have just used the custom domain during extraction, but this looks acceptable too.

I guess I would have liked something less (overtly) regexp-based as well, but it’s not like I have a better solution at the ready.

File: yt_dlp/extractor/substack.py (hunk 4 above)
Comment (elyse0) [reply yt_dlp/extractor/substack.py]: I think uploader refers to the channel/newsletter/blog where it was uploaded, not the author 🤔 

I've improved it a bit in https://github.com/yt-dlp/yt-dlp/pull/4011/commits/f9ab764f2530cd6ea44189cfd02da855d94af24e, what do you think?

File: yt_dlp/extractor/generic.py [hunk 5]
```diff
@@ -3037,6 +3064,11 @@ def _real_extract(self, url):
             # Don't set the extractor because it can be a track url or an album
             return self.url_result(burl)
 
+        # Check for Substack custom domains
+        substack_url = SubstackIE._extract_url(webpage, url)
+        if substack_url:
+            return self.url_result(substack_url, SubstackIE.ie_key())
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3070]:
```suggestion
            return self.url_result(substack_url, SubstackIE)
```

File: yt_dlp/extractor/substack.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
+import re
+from urllib.parse import urlparse
+
+from .common import InfoExtractor
+from ..utils import (
+    traverse_obj,
+    str_or_none,
+)
```
Comment (pukkandan) [inline yt_dlp/extractor/substack.py:8]:
```suggestion
import urllib.parse

from .common import InfoExtractor
from ..utils import (
    str_or_none,
    traverse_obj,
)
```

File: yt_dlp/extractor/substack.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    }]
+
+    @classmethod
+    def _extract_url(cls, webpage, url):
+        if not re.search(r'<script[^>]+src=["\']https://substackcdn.com/[^"\']+\.js', webpage):
+            return
+
+        subdomain = re.search(r'{[^}]*["\']subdomain["\']\s*:\s*["\'](?P<subdomain>[^"]+)', webpage).group('subdomain')
+        parsed = urlparse(url)
+        return parsed._replace(netloc=f'{subdomain}.substack.com').geturl()
```
Comment (pukkandan) [inline yt_dlp/extractor/substack.py:58]:
```suggestion
        mobj = re.search(r'{[^}]*["\']subdomain["\']\s*:\s*["\'](?P<subdomain>[^"]+)', webpage)
        if mobj:
            parsed = urlparse(url)
            return parsed._replace(netloc=f'{mobj.group("subdomain")}.substack.com').geturl()
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/substack.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
+import re
+import urllib.parse
+
+from .common import InfoExtractor
+from ..utils import (
+    str_or_none,
+    traverse_obj,
+)
```
Comment (pukkandan) [inline yt_dlp/extractor/substack.py:8]:
```suggestion
from ..utils import str_or_none, traverse_obj
```

File: yt_dlp/extractor/substack.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        self._sort_formats(formats)
+        return {
+            'id': str(webpage_info['post']['id']),
+            'formats': formats,
+            'subtitles': subtitles,
+            'title': traverse_obj(webpage_info, ('post', 'title')),
+            'description': traverse_obj(webpage_info, ('post', 'description')),
+            'thumbnail': traverse_obj(webpage_info, ('post', 'cover_image')),
+            'uploader': traverse_obj(webpage_info, ('pub', 'name')),
+            'uploader_id': str_or_none(traverse_obj(webpage_info, ('pub', 'author_id'))),
```
Comment (fstirlitz) [inline yt_dlp/extractor/substack.py:99]: `.pub.name` identifies the newsletter, but `.pub.author_id` identifies the author. If you want to have ‘uploader’ mean ‘newsletter’ for this extractor, you should use `.pub.id` or `.post.publication_id` as `uploader_id`.

Also take note that some newsletters may have multiple authors. <https://www.mollymovieclub.com/p/interstellar> names two authors, but this extracts only one of them.

I remember there being some (proposal to add?) ‘crew’ credits metadata fields, maybe post authors (`.publishedBylines`) could be squeezed there.

Comment (elyse0) [reply yt_dlp/extractor/substack.py]: Should be good now

The closest thing that I found in some extractors is `cast`, could it work?

Comment (fstirlitz) [reply yt_dlp/extractor/substack.py]: Maybe. But I don’t insist on it.

**Pull request merged**